docs/api-conventions.md

```diff
@@ -239,37 +229,178 @@ Examples:
 TODO: Plugins, extensions, nested kinds, headers
 
 
-Status codes
-------------
-
-The following status codes may be returned by the API.
+HTTP Status codes
+-----------------
 
-TODO: Document when each of these codes is returned
+The following HTTP status codes may be returned by the API.
 
 #### Success codes
 
-* `StatusOK`
-* `StatusCreated`
-* `StatusAccepted`
-* `StatusNoContent`
+* `200 StatusOK`
+  * Indicates that the request completed succesfully.
+* `201 StatusCreated`
+  * Indicates that the request to create kind completed succesfully.
+* `204 StatusNoContent`
+  * Indicates that the request completed succesfully, and the response contains no body.
+  * Returned in response to HTTP OPTIONS requests.
 
 #### Error codes
+* `307 StatusTemporaryRedirect`
+  * Indicates that the address for the requested resource has changed.
+  * Suggested client recovery behavior
+    * Follow the redirect.
+* `400 StatusBadRequest`
+  * Indicates the requested is invalid.
+  * Suggested client recovery behavior:
+    * Do not retry. Fix the request.
+* `403 StatusForbidden`
+  * Indicates that the server can be reached and understood the request, but refuses to take any further action, because it is configured to deny access for some reason to the requested resource by the client.
+  * Suggested client recovery behavior
+    * Do not retry. Fix the request.
+* `404 StatusNotFound`
+  * Indicates that the requested resource does not exist.
+  * Suggested client recovery behavior
+    * Do not retry. Fix the request.
+* `405 StatusMethodNotAllowed`
+  * Indicates that that the action the client attempted to perform on the resource was not supported by the code.
+  * Suggested client recovery behavior
+    * Do not retry. Fix the request.
+* `409 StatusConflict`
+  * Indicates that an item with the given identifier already exists and can't be created or can't be updated.
+  * Suggested client recovery behavior
+    * Do not retry. Fix the request: if creating a new item, change the identifier. If updating an item, ensure the `ResourceVersion` is up to date.
```

Comment:
This case may require more explanation. You may also want to add a forward reference to the more detailed status response, below.

Unlike the cases above, an automatic response to 409 should typically be feasible. 

For example, when updating, usually the client would GET the object, check the client's preconditions, re-merge changes if still valid according to preconditions, and retry with the updated request (including resourceVersion).

When creating, the client would GET and compare that the fields they tried to set matched the created object. Name uniqueness ensures idempotence in the presence of failure to receive a response, including due to failure of the client. A pre-existing object should be considered ok, provided it matched what the client wanted.